### ISSUE TITLE
Cross compile for win-x64 fails on Xubuntu:  "[38/54]: Building all-gcc... error"

### ISSUE THREAD

**Issue by ghost** [rdp/ffmpeg-windows-build-helpers#195]
In file included from ./tm.h:23:0,
                 from ../../source/gcc-4.9.3/gcc/cp/friend.c:23:
../../source/gcc-4.9.3/gcc/config/i386/mingw32.h:139:3: warning: invalid suffix on literal; C++11 requires a space between literal and string macro [-Wliteral-suffix]
   "%{mthreads:-lmingwthrd} -lmingw32 \
   ^
In file included from ../../source/gcc-4.9.3/gcc/cp/except.c:1013:0:
cfns.gperf: In function ‘const char* libc_name_p(const char*, unsigned int)’:
cfns.gperf:101:1: error: ‘const char* libc_name_p(const char*, unsigned int)’ redeclared inline with ‘gnu_inline’ attribute
cfns.gperf:26:14: note: ‘const char* libc_name_p(const char*, unsigned int)’ previously declared here
cfns.gperf: At global scope:
cfns.gperf:26:14: warning: inline function ‘const char* libc_name_p(const char*, unsigned int)’ used but never defined
Makefile:1058: recipe for target 'cp/except.o' failed
make[1]: *** [cp/except.o] Error 1
make[1]: *** Waiting for unfinished jobs....
rm gcc.pod
make[1]: Leaving directory '/home/ran/ffmpeg-windows-build-helpers/sandbox/cross_compilers/pkgs/gcc/build/gcc'
Makefile:3968: recipe for target 'all-gcc' failed
make: *** [all-gcc] Error 2

**Comment by rdp** [comment]
what is your version of gcc?

**Comment by alice-margatroid** [comment]
Thanks for the reply, I'm using:
$ apt-show-versions gcc
gcc:amd64/yakkety 4:6.1.1-1ubuntu2 uptodate

**Comment by rdp** [comment]
could you try it again to see if it was a fluke with mingw-w64 git?

**Comment by alice-margatroid** [comment]
Are 4 tries on a freshly set-up Xubuntu enough?

**Comment by alice-margatroid** [comment]
Don't mistake my last post with sarcasm. I am still awaiting this issue to be resolved.

At least provide me with information about which distro to use out of the box and compile without issues and close it. What gcc do you use and do your own cross-compiles on GNU/Linux still work?

**Comment by rdp** [comment]
It is to see if it was a fluke that mingw-w64 git has since fixed (since
it's working fine here...). At worst if it still fails I've heard of people
newing up ubuntu VM's FWIW.

On Thu, Jan 12, 2017 at 10:32 PM, alice-margatroid <notifications@github.com

**Comment by rzcat** [comment]
Seems same error with mine, closed but unsolved. I ended up using Windows instead.
https://github.com/rdp/ffmpeg-windows-build-helpers/issues/181
To add up, I'm also using LXDE on Fedora, and my gcc version is
	gcc (GCC) 6.3.1 20161221 (Red Hat 6.3.1-1)

**Comment by alice-margatroid** [comment]
@rzcat 
Thanks, will have to try that. Compiling on *NIX is just so much comfier.

@rdp 
>At worst if it still fails I've heard of people newing up ubuntu VM's FWIW.
I must be semantically misunderstanding that sentence, since yakkety is the latest stable version you can have until April. My own cross-compiles with less encoders were all successful, but won't get me anywhere. So I gave your repo a shot. Why do you refuse to disclose your distro? It's normal to talk about that.

**Comment by rdp** [comment]
16.04 with gcc 5.4 works.
It's possible newer versions of gcc refuse to compile the gcc it uses (4.9.3 I think...)

**Comment by alice-margatroid** [comment]
@rdp 
That went absolutely smooth for me when compiling on the old version. Thanks for the suggestion and cheers!

**Comment by msmollin** [comment]
Just to add my own experience, I spooled up a stock Ubuntu 16.04.1 LTS Server instance in VirtualBox and had no issues starting up the compilation (it's still compiling I post this, but it's made it beyond all the early setup). I did do a `dist-upgrade` on it to make sure I was on the latest kernel. So whatever version of GCC gets shipped with 16.04 LTS seems to be good, as @rdp and @alice-margatroid have also mentioned.

**Comment by rdp** [comment]
OK appears the problem is with GCC 6.x with no known work around yet, added
a note to the READMe anyway...

**Comment by Darkmasterhk** [comment]
it's been a while since I tried this the last time but I was able to compile ffmpeg with your script on Ubuntu 16.10 (gcc 6) after changing your gcc version from 4.9.3 to 4.9.4 (inside the script)

**Comment by benurb** [comment]
Yep, can confirm that it works with 4.9.4. For anyone willing to try that out.
1. Change version in cross_compile_ffmpeg.sh
```diff
@@ -179,7 +179,7 @@ install_cross_compiler() {
     local zeranoe_script_name=mingw-w64-build-3.6.7.local
     # add --mingw-w64-ver=git for updated tuner.h [dshow dtv] at least not present in 4.0.6 TODO bump to v 5 when released, if released
     # actually git make "faster" builds for some reason, so leave for now, known working commit: d9ce1abe40efb835609e646b1533acab4a404d03
-    local zeranoe_script_options="--clean-build --disable-shared --default-configure  --pthreads-w32-ver=2-9-1 --cpu-count=$gcc_cpu_count --mingw-w64-ver=git --gcc-ver=4.9.3"
+    local zeranoe_script_options="--clean-build --disable-shared --default-configure  --pthreads-w32-ver=2-9-1 --cpu-count=$gcc_cpu_count --mingw-w64-ver=git --gcc-ver=4.9.4"
     if [[ ($compiler_flavors == "win32" || $compiler_flavors == "multi") && ! -f ../$win32_gcc ]]; then
       echo "building win32 cross compiler..."
       download_gcc_build_script $zeranoe_script_name```

2. Run ./cross_compile_ffmpeg.sh once. It will quit with an error, that this version is not allowed.
3. Comment out the download code
```diff
@@ -191,7 +191,7 @@ install_cross_compiler() {
     fi
     if [[ ($compiler_flavors == "win64" || $compiler_flavors == "multi") && ! -f ../$win64_gcc ]]; then
       echo "building win64 x86_64 cross compiler..."
-      download_gcc_build_script $zeranoe_script_name
+      #download_gcc_build_script $zeranoe_script_name
       nice ./$zeranoe_script_name $zeranoe_script_options --build-type=win64 || exit 1
       if [[ ! -f ../$win64_gcc ]]; then
         echo "failure building 64 bit gcc? recommend nuke sandbox (rm -rf sandbox) and start over..."
```
4. Change `gcc_old_release_ver` from 4.9.3 to 4.9.4 in `sandbox/cross_compilers/mingw-w64-build-3.6.7.local`
5. Run ./cross_compile_ffmpeg.sh again. It should work now

**Comment by rdp** [comment]
OK bumped it to 4.9.4 HTH

**Comment by rzcat** [comment]
Finally got time to test, my environment is Fedora 25 x86_64(gcc 6.3.1), and the cmd line options for the script is
`--disable-nonfree=y --high-bitdepth=y --compiler-flavors=win64 --build-lsw=y --prefer-stable=n`
Sucess, and the result for libx265 is 
`x265 2.3+17-6e348252e902:[Windows][GCC 4.9.4][64 bit] 10bit`
and I did try to modify the line 182 changing the **--gcc-ver=4.9.3** to **--gcc-ver=6.3.0** only to get the same error:
`[33/49]: Building all-gcc... error`
and the final part of build.log shows
```shell
GIMPLE decision tree has 1428 leafs, maximum depth 10 and a total number of 5495 nodes
removed 734 duplicate tails
build/genmatch --generic ../../source/gcc-6.3.0/gcc/match.pd \
    > tmp-generic-match.c
GENERIC decision tree has 1428 leafs, maximum depth 10 and a total number of 5495 nodes
removed 734 duplicate tails
/bin/sh ../../source/gcc-6.3.0/gcc/../move-if-change tmp-gimple-match.c \
    					gimple-match.c
/bin/sh ../../source/gcc-6.3.0/gcc/../move-if-change tmp-generic-match.c \
    					generic-match.c
echo timestamp > s-match
g++ -fno-PIE -c   -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings  -Wno-unused -DHAVE_CONFIG_H -I. -I. -I../../source/gcc-6.3.0/gcc -I../../source/gcc-6.3.0/gcc/. -I../../source/gcc-6.3.0/gcc/../include -I../../source/gcc-6.3.0/gcc/../libcpp/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/gmp/gmp-6.1.1-x86_64/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/mpfr/mpfr-3.1.3-x86_64/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/mpc/mpc-1.0.3-x86_64/include  -I../../source/gcc-6.3.0/gcc/../libdecnumber -I../../source/gcc-6.3.0/gcc/../libdecnumber/bid -I../libdecnumber -I../../source/gcc-6.3.0/gcc/../libbacktrace -I/ff_comp_test/sandbox/cross_compilers/pkgs/isl/isl-0.16.1-x86_64/include  -o gimple-match.o -MT gimple-match.o -MMD -MP -MF ./.deps/gimple-match.TPo gimple-match.c
g++ -fno-PIE -c   -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings  -Wno-unused -DHAVE_CONFIG_H -I. -I. -I../../source/gcc-6.3.0/gcc -I../../source/gcc-6.3.0/gcc/. -I../../source/gcc-6.3.0/gcc/../include -I../../source/gcc-6.3.0/gcc/../libcpp/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/gmp/gmp-6.1.1-x86_64/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/mpfr/mpfr-3.1.3-x86_64/include -I/ff_comp_test/sandbox/cross_compilers/pkgs/mpc/mpc-1.0.3-x86_64/include  -I../../source/gcc-6.3.0/gcc/../libdecnumber -I../../source/gcc-6.3.0/gcc/../libdecnumber/bid -I../libdecnumber -I../../source/gcc-6.3.0/gcc/../libbacktrace -I/ff_comp_test/sandbox/cross_compilers/pkgs/isl/isl-0.16.1-x86_64/include  -o generic-match.o -MT generic-match.o -MMD -MP -MF ./.deps/generic-match.TPo generic-match.c
g++ -c   -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -DGENERATOR_FILE -fno-PIE -I. -Ibuild -I../../source/gcc-6.3.0/gcc -I../../source/gcc-6.3.0/gcc/build -I../../source/gcc-6.3.0/gcc/../include  -I../../source/gcc-6.3.0/gcc/../libcpp/include  \
	-o build/genattrtab.o ../../source/gcc-6.3.0/gcc/genattrtab.c
g++   -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -DGENERATOR_FILE -fno-PIE  -no-pie -o build/genattrtab \
    build/genattrtab.o build/rtl.o build/read-rtl.o build/ggc-none.o build/vec.o build/min-insn-modes.o build/gensupport.o build/print-rtl.o build/hash-table.o build/read-md.o build/errors.o ../build-x86_64-pc-linux-gnu/libiberty/libiberty.a
build/genattrtab ../../source/gcc-6.3.0/gcc/common.md ../../source/gcc-6.3.0/gcc/config/i386/i386.md insn-conditions.md \
	-Atmp-attrtab.c -Dtmp-dfatab.c -Ltmp-latencytab.c
Makefile:2201: recipe for target 's-attrtab' failed
make[1]: *** [s-attrtab] Killed
make[1]: Leaving directory '/ff_comp_test/sandbox/cross_compilers/pkgs/gcc/build/gcc'
Makefile:4130: recipe for target 'all-gcc' failed
make: *** [all-gcc] Error 2
```

**Comment by rdp** [comment]
hopefully fixed by moving to gcc 4.9.4 now

**Comment by rdp** [comment]
OK things have changed quite  bit, let me know if still failing...
